tests/example.c

```diff
@@ -21,7 +21,7 @@ void add_user(int user_id, char *name)
         s->id = user_id;
         HASH_ADD_INT( users, id, s );  /* id: name of key field */
     }
-    strcpy(s->name, name);
+    strncpy(s->name, name, sizeof(s->name));
```

Comment:
`strcpy` is correct — or at least, it doesn't hide the bad behavior if `name` is too long for the `s->name` field.

`strncpy` would avoid the buffer overflow on long names, but only by truncating the entered `name` — and incidentally, if the `name` were shorter than `sizeof(s->name)`, it'd pad with null bytes (which is harmless but inefficient).

The way to make this code "safe" would be to slap an `assert(strlen(name) < sizeof(s->name))` on the line above... but as this is just a toy example, I don't know that that's warranted, either.

Reply:
> `strcpy` is correct — or at least, it doesn't hide the bad behavior if `name` is too long for the `s->name` field.

It is neither correct nor does it give the user any useful information why it just crashed. And leaving an exploitable buffer overflow there for cosmetic reasons is not a sign of code quality.

> `strncpy` would avoid the buffer overflow on long names, but only by truncating the entered `name` — and incidentally, if the `name` were shorter than `sizeof(s->name)`, it'd pad with null bytes (which is harmless but inefficient).
> 
> The way to make this code "safe" would be to slap an `assert(strlen(name) < sizeof(s->name))` on the line above... but as this is just a toy example, I don't know that that's warranted, either.

I would not use assertions in that situation but it gives at least some information why this application stops doing what it should do.

Reply:
> `strncpy` would avoid the buffer overflow on long names, but only by truncating the entered `name` — and incidentally, if the `name` were shorter than `sizeof(s->name)`, it'd pad with null bytes (which is harmless but inefficient).

By the way: There is no NUL padding, just one trailing NUL byte. See `man 3 strncpy`.

Reply:
Today you learned. :) https://linux.die.net/man/3/strncpy

Reply:
Agreed. At least one useful outcome from this discussion.